examples/Superresolution/enet-pat.py

```diff
@@ -0,0 +1,334 @@
+                    x = Conv2D('conv4b', x, 512, stride=2)
+
+                x = FullyConnected('fc0', x, 1024, nl=LeakyReLU)
+                x = FullyConnected('fc1', x, 1, nl=tf.identity)
+            return x
+
+        def additional_losses(a, b):
+            with tf.variable_scope('VGG19'):
+                x = tf.concat([a, b], axis=0)
+                x = tf.reshape(x, [2 * BATCH_SIZE, 128, 128, 3])
+                # VGG 19
+                with argscope(Conv2D, kernel_shape=3, nl=tf.nn.relu):
+                    conv1_1 = Conv2D('conv1_1', x, 64)
+                    conv1_2 = Conv2D('conv1_2', conv1_1, 64)
+                    pool1 = MaxPooling('pool1', conv1_2, 2)  # 64
+                    conv2_1 = Conv2D('conv2_1', pool1, 128)
+                    conv2_2 = Conv2D('conv2_2', conv2_1, 128)
+                    pool2 = MaxPooling('pool2', conv2_2, 2)  # 32
+                    conv3_1 = Conv2D('conv3_1', pool2, 256)
+                    conv3_2 = Conv2D('conv3_2', conv3_1, 256)
+                    conv3_3 = Conv2D('conv3_3', conv3_2, 256)
+                    conv3_4 = Conv2D('conv3_4', conv3_3, 256)
+                    pool3 = MaxPooling('pool3', conv3_4, 2)  # 16
+                    conv4_1 = Conv2D('conv4_1', pool3, 512)
+                    conv4_2 = Conv2D('conv4_2', conv4_1, 512)
+                    conv4_3 = Conv2D('conv4_3', conv4_2, 512)
+                    conv4_4 = Conv2D('conv4_4', conv4_3, 512)
+                    pool4 = MaxPooling('pool4', conv4_4, 2)  # 8
+                    conv5_1 = Conv2D('conv5_1', pool4, 512)
+                    conv5_2 = Conv2D('conv5_2', conv5_1, 512)
+                    conv5_3 = Conv2D('conv5_3', conv5_2, 512)
+                    conv5_4 = Conv2D('conv5_4', conv5_3, 512)
+                    pool5 = MaxPooling('pool5', conv5_4, 2)  # 4
+
+            # perceptual loss
+            with tf.name_scope('perceptual_loss'):
+                phi_a_1, phi_b_1 = tf.split(pool2, 2, axis=0)
+                phi_a_2, phi_b_2 = tf.split(pool5, 2, axis=0)
+
+                logger.info('create perceptual loss for layer {} with shape {}'.format(pool2.name, pool2.get_shape()))
+                pool2_loss = mse(phi_a_1, phi_b_1, name='perceptual_loss_pool2')
+                logger.info('create perceptual loss for layer {} with shape {}'.format(pool5.name, pool5.get_shape()))
+                pool5_loss = mse(phi_a_2, phi_b_2, name='perceptual_loss_pool5')
+
+            # texture loss
+            with tf.name_scope('texture_loss'):
+                def texture_loss(x, p=16):
+                    x = normalize(x)
+                    _, h, w, _ = x.get_shape().as_list()
+                    rsl = []
+                    logger.info('create texture loss for layer {} with shape {}'.format(
+                        x.name, x.get_shape()))
+                    for i in range(0, h - p, p - 1):
+                        for j in range(0, w - p, p - 1):
```

Comment:
tf.space_to_batch_nd + reshape ?

Reply:
I feel more comfortable having the loop, which is IMHO much easier to parse than hiding this behind some functions. But I am fine with `tf.space_to_batch_nd` if this is equivalent.

Reply:
Perhaps. Just feel looping over pixels isn't very tensorflowish.
```python
x = tf.space_to_batch_nd(x, [p, p], [[0,0],[0,0]])
x = tf.reshape(x, [p, p, -1, h, w, c])
x = tf.transpose(x, [2, 0, 1, 3, 4, 5])
patches_a, patches_b = tf.split(x, 2)   # each is b,p,p,h/p,w/p,c
```

Reply:
Spiced with a pinch of pytorch style. Honestly, I was not aware of `tf.space_to_batch_nd`.

Reply:
fix:
```python
x = tf.space_to_batch_nd(x, [p, p], [[0,0],[0,0]])
x = tf.reshape(x, [p, p, -1, h/p, w/p, c])
x = tf.transpose(x, [2, 3, 4, 0, 1, 5])
patches_a, patches_b = tf.split(x, 2)   # each is b,h/p,w/p,p,p,c
```

Reply:
Haven't changed it, but then
```python
x = tf.space_to_batch_nd(x, [p, p], [[0,0],[0,0]])
x = tf.reshape(x, [p, p, -1, h/p, w/p, c])
x = tf.transpose(x, [2, 3, 4, 0, 1, 5])
patches_a, patches_b = tf.split(x, 2)   # each is b,h/p,w/p,p,p,c

patches_a = tf.reshape(patches_a, [-1, p, p, c])
patches_b = tf.reshape(patches_b, [-1, p, p, c])
```
should work